gdx/src/com/badlogic/gdx/graphics/g3d/shadow/system/realistic/RealisticShadowSystem.java

```diff
@@ -0,0 +1,451 @@
+		if( spotCameras.containsKey(spot) ) {
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public boolean hasLight (DirectionalLight dir) {
+		if( dirCameras.containsKey(dir) ) {
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public boolean hasLight (PointLight point) {
+		if( pointCameras.containsKey(point) ) {
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public void update() {
+		for(ObjectMap.Entry<SpotLight, LightProperties> e : spotCameras) {
+			e.value.camera.position.set(e.key.position);
+			e.value.camera.direction.set(e.key.direction);
+			Vector2 nearFar = nearFarAnalyzer.analyze(e.key, e.value.camera);
+			e.value.camera.near = nearFar.x;
+			e.value.camera.far = nearFar.y;
+			e.value.camera.update();
+		}
+
+		for(ObjectMap.Entry<DirectionalLight, LightProperties> e : dirCameras) {
+			e.value.camera.direction.set(e.key.direction);
+			directionalAnalyzer
+			.analyze(
+				e.key,
+				scene.getCamera().frustum,
+				e.value.camera.direction)
+				.set(e.value.camera);
+			e.value.camera.update();
+		}
+
+		for(ObjectMap.Entry<PointLight, PointLightProperties> e : pointCameras) {
+			for(ObjectMap.Entry<CubemapSide, LightProperties> c : e.value.properties) {
+				c.value.camera.position.set(e.key.position);
+				Vector2 nearFar = nearFarAnalyzer.analyze(e.key, c.value.camera);
+				c.value.camera.near = nearFar.x;
+				c.value.camera.far = nearFar.y;
+				c.value.camera.update();
+			}
+		}
+	}
+
+	@Override
+	public void begin (int n) {
+		spotCameraIterator = spotCameras.iterator();
+		spotCameraIterator.reset();
+
+		dirCameraIterator = dirCameras.iterator();
+		dirCameraIterator.reset();
+
+		pointCameraIterator = pointCameras.iterator();
+		pointCameraIterator.reset();
+
+		currentPointSide = 6;
+		currentPass = n;
+
+		allocator.begin();
+		frameBuffer.begin();
+		Gdx.gl.glClearColor(1, 1, 1, 1);
+		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT | GL20.GL_DEPTH_BUFFER_BIT);
+		Gdx.gl.glEnable(GL20.GL_SCISSOR_TEST);
+	}
+
+	@Override
+	public Camera next () {
+		Camera camera = nextDirectional();
+		if( camera != null )
+			return camera;
+
+		camera = nextSpot();
+		if( camera != null )
+			return camera;
+
+		camera = nextPoint();
+		if( camera != null )
+			return camera;
+
+		return null;
+	}
+
+	protected Camera nextDirectional() {
+		if( !dirCameraIterator.hasNext() ) {
+			return null;
+		}
+
+		LightProperties lp = dirCameraIterator.next().value;
+		processViewport(lp);
+		return lp.camera;
+	}
+
+	protected Camera nextSpot() {
+		if( !spotCameraIterator.hasNext() ) {
+			return null;
+		}
+
+		LightProperties lp = spotCameraIterator.next().value;
+		if( !lightFilter.filter(currentPass, spotCameras.findKey(lp, true), lp.camera) ) {
+			return nextSpot();
+		}
+		processViewportCamera(lp.camera, processViewport(lp));
+		return lp.camera;
+	}
+
+	protected Camera nextPoint() {
+		if( !pointCameraIterator.hasNext() && currentPointSide > 5 ) {
+			return null;
+		}
+
+		if( currentPointSide > 5 )
```

Comment:
I do it manually because I don't know why my formatter doesn't work... It's long...